### ISSUE TITLE
 Low mAP on pytorch 1.0 branch

### ISSUE THREAD

**Issue by DHZS** [jwyang/faster-rcnn.pytorch#419]
I downloaded the trained model from here.

model | #GPUs | batch size | lr | lr_decay | max_epoch | time/epoch | mem/GPU | mAP
-- | -- | -- | -- | -- | -- | -- | -- | --
VGG-16 | 1 | 1 | 1e-3 | 5 | 6 | 0.76 hr | 3265MB | 70.1

I use pytorch 1.0, and the mAP on pascal voc 2007 is 65.8, which is lower than 70.1.
Here is the APs for each category.
```
VOC07 metric? Yes
AP for aeroplane = 0.6559
AP for bicycle = 0.7750
AP for bird = 0.6105
AP for boat = 0.4600
AP for bottle = 0.4542
AP for bus = 0.7689
AP for car = 0.7686
AP for cat = 0.8393
AP for chair = 0.4700
AP for cow = 0.6777
AP for diningtable = 0.6060
AP for dog = 0.7668
AP for horse = 0.8110
AP for motorbike = 0.6922
AP for person = 0.7295
AP for pottedplant = 0.3714
AP for sheep = 0.6060
AP for sofa = 0.6509
AP for train = 0.7408
AP for tvmonitor = 0.7092
Mean AP = 0.6582
```

I also tested on the master branch using pytorch 0.4.0. The results are as follows:
```
VOC07 metric? Yes
AP for aeroplane = 0.7055
AP for bicycle = 0.7761
AP for bird = 0.6638
AP for boat = 0.5457
AP for bottle = 0.5221
AP for bus = 0.8086
AP for car = 0.8450
AP for cat = 0.8411
AP for chair = 0.5025
AP for cow = 0.7824
AP for diningtable = 0.6539
AP for dog = 0.7754
AP for horse = 0.8294
AP for motorbike = 0.7267
AP for person = 0.7726
AP for pottedplant = 0.4223
AP for sheep = 0.7188
AP for sofa = 0.6555
AP for train = 0.7526
AP for tvmonitor = 0.7277
Mean AP = 0.7014
```

Is there something wrong about the evaluation code?

**Comment by DHZS** [comment]
I found that the values `pooled_feat` returned by `self.RCNN_roi_align(base_feat, rois.view(-1, 5))` in pytorch 0.4.0 and 1.0 are different. So, which RoIAlign implementation is the right one?

**Comment by xgmiao** [comment]
I think maybe the pytorch0.4 RoIAlign implementation is right.

**Comment by adityaarun1** [comment]
Hi,

For the PyTorch-1.0 branch, I re-ran the code and it gives me 70.26% mAP when using VGG-16 and 74.90% using ResNet-101. These numbers are achieved using the default parameter values on a single GPU. These numbers are consistent with the ones reported by @jwyang in the readme. Here are the details:

VGG-16 (7th epoch)
```
VOC07 metric? Yes
AP for aeroplane = 0.7071
AP for bicycle = 0.7714
AP for bird = 0.6922
AP for boat = 0.5591
AP for bottle = 0.5510
AP for bus = 0.7913
AP for car = 0.8458
AP for cat = 0.8173
AP for chair = 0.4943
AP for cow = 0.7680
AP for diningtable = 0.6090
AP for dog = 0.8058
AP for horse = 0.8283
AP for motorbike = 0.7468
AP for person = 0.7788
AP for pottedplant = 0.4439
AP for sheep = 0.7058
AP for sofa = 0.6440
AP for train = 0.7584
AP for tvmonitor = 0.7345
Mean AP = 0.7026
```

ResNet-101 (7th epoch)
```
VOC07 metric? Yes
AP for aeroplane = 0.7883
AP for bicycle = 0.8002
AP for bird = 0.7752
AP for boat = 0.6471
AP for bottle = 0.6047
AP for bus = 0.8135
AP for car = 0.8585
AP for cat = 0.8698
AP for chair = 0.5443
AP for cow = 0.8122
AP for diningtable = 0.6728
AP for dog = 0.8568
AP for horse = 0.8484
AP for motorbike = 0.7912
AP for person = 0.7880
AP for pottedplant = 0.4702
AP for sheep = 0.7661
AP for sofa = 0.7292
AP for train = 0.7955
AP for tvmonitor = 0.7487
Mean AP = 0.7490
```

**Comment by DHZS** [comment]
Thank you very much, @adityaarun1.
I saw you reply at https://github.com/ruotianluo/pytorch-faster-rcnn/pull/122. I trained a model from scratch and got 70.26% mAP. It seems that it is important to retrain the model in pytorch 1.0.
Thanks again, you have helped me a lot.

**Comment by zc-alexfan** [comment]
I obtained mAP 65.8 using VGG on PASCAL_VOC. However, for Visual Genome, my mAP is always close to 0. 

I am wondering if anyone has an mAP of 4.4, which is demonstrated in the bench mark?

**Comment by adityaarun1** [comment]
I believe you are required to train all the models you need from scratch. The pre-trained models were trained for PyTorch <= 0.4.0, and there has been breaking changes since.

**Comment by zc-alexfan** [comment]
I see. So if I switch to Pytorch 0.4. It should work? Also, have you verified whether the visual genome pretrained model works as promised?

**Comment by adityaarun1** [comment]
Yes, switching to Pytorch=0.4.0 should work.
I haven't checked Visual Genome pretrained model personally. But I see no reason why it should not work for Pytorch=0.4.0

**Comment by zc-alexfan** [comment]
Thanks for the quick reply! It seems that [setting up this repo for pytorch 0.4 doesn't work anymore. ](https://github.com/pytorch/pytorch/issues/15645)

I also noticed your repo (https://github.com/adityaarun1/Detectron.pytorch). 

I am setting up a FasterRCNN model that takes an image and outputs rois, roi probability over labels, and the pooled features corresponding to the rois. 

Do you suggest to use **Detection** or this **fasterRCNN** repo to obtain the above information? Because I have been stuck in the mAP problem for weeks and think about pivoting to `Detectron `.

**Comment by adityaarun1** [comment]
Make sure you are on Pytorch `0.4.0` and not on `0.4.1`, and that you are on the correct branch. In your case, either of the repositories should be fine.

**Comment by zc-alexfan** [comment]
Hi, I have setup the repo using `pytorch 0.4.0` on the `master` branch. The mAP I obtained is still very low. I have attached my config file and other print out from the program. I am hoping if I could ask for a sanity check of my configuration? Thank you!

**Comment by zc-alexfan** [comment]
Command to run: 
~~~
python3.6 test_net.py --dataset vg --net vgg16 \
                      --checksession 1 --checkepoch 19 --checkpoint 48611 \
                      --cuda
~~~

Print out: 
~~~
Called with args:
Namespace(cfg_file='cfgs/vgg16.yml', checkepoch=19, checkpoint=48611, checksession=1, class_agnostic=False, cuda=True, dataset='vg', large_scale=False, load_dir='models', mGPUs=False, net='vgg16', parallel_type=0, set_cfgs=None, vis=False)
Using config:
{'ANCHOR_RATIOS': [0.5, 1, 2],
 'ANCHOR_SCALES': [4, 8, 16, 32],
 'CROP_RESIZE_WITH_MAX_POOL': False,
 'CUDA': False,
 'DATA_DIR': '/home/alex/faster-rcnn.pytorch/data',
 'DEDUP_BOXES': 0.0625,
 'EPS': 1e-14,
 'EXP_DIR': 'vgg16',
 'FEAT_STRIDE': [16],
 'GPU_ID': 0,
 'MATLAB': 'matlab',
 'MAX_NUM_GT_BOXES': 20,
 'MOBILENET': {'DEPTH_MULTIPLIER': 1.0,
               'FIXED_LAYERS': 5,
               'REGU_DEPTH': False,
               'WEIGHT_DECAY': 4e-05},
 'PIXEL_MEANS': array([[[102.9801, 115.9465, 122.7717]]]),
 'POOLING_MODE': 'align',
 'POOLING_SIZE': 7,
 'RESNET': {'FIXED_BLOCKS': 1, 'MAX_POOL': False},
 'RNG_SEED': 3,
 'ROOT_DIR': '/home/alex/faster-rcnn.pytorch',
 'TEST': {'BBOX_REG': True,
          'HAS_RPN': True,
          'MAX_SIZE': 1000,
          'MODE': 'nms',
          'NMS': 0.3,
          'PROPOSAL_METHOD': 'gt',
          'RPN_MIN_SIZE': 16,
          'RPN_NMS_THRESH': 0.7,
          'RPN_POST_NMS_TOP_N': 300,
          'RPN_PRE_NMS_TOP_N': 6000,
          'RPN_TOP_N': 5000,
          'SCALES': [600],
          'SVM': False},
 'TRAIN': {'ASPECT_GROUPING': False,
           'BATCH_SIZE': 256,
           'BBOX_INSIDE_WEIGHTS': [1.0, 1.0, 1.0, 1.0],
           'BBOX_NORMALIZE_MEANS': [0.0, 0.0, 0.0, 0.0],
           'BBOX_NORMALIZE_STDS': [0.1, 0.1, 0.2, 0.2],
           'BBOX_NORMALIZE_TARGETS': True,
           'BBOX_NORMALIZE_TARGETS_PRECOMPUTED': True,
           'BBOX_REG': True,
           'BBOX_THRESH': 0.5,
           'BG_THRESH_HI': 0.5,
           'BG_THRESH_LO': 0.0,
           'BIAS_DECAY': False,
           'BN_TRAIN': False,
           'DISPLAY': 10,
           'DOUBLE_BIAS': True,
           'FG_FRACTION': 0.25,
           'FG_THRESH': 0.5,
           'GAMMA': 0.1,
           'HAS_RPN': True,
           'IMS_PER_BATCH': 1,
           'LEARNING_RATE': 0.01,
           'MAX_SIZE': 1000,
           'MOMENTUM': 0.9,
           'PROPOSAL_METHOD': 'gt',
           'RPN_BATCHSIZE': 256,
           'RPN_BBOX_INSIDE_WEIGHTS': [1.0, 1.0, 1.0, 1.0],
           'RPN_CLOBBER_POSITIVES': False,
           'RPN_FG_FRACTION': 0.5,
           'RPN_MIN_SIZE': 8,
           'RPN_NEGATIVE_OVERLAP': 0.3,
           'RPN_NMS_THRESH': 0.7,
           'RPN_POSITIVE_OVERLAP': 0.7,
           'RPN_POSITIVE_WEIGHT': -1.0,
           'RPN_POST_NMS_TOP_N': 2000,
           'RPN_PRE_NMS_TOP_N': 12000,
           'SCALES': [600],
           'SNAPSHOT_ITERS': 5000,
           'SNAPSHOT_KEPT': 3,
           'SNAPSHOT_PREFIX': 'res101_faster_rcnn',
           'STEPSIZE': [30000],
           'SUMMARY_INTERVAL': 180,
           'TRIM_HEIGHT': 600,
           'TRIM_WIDTH': 600,
           'TRUNCATED': False,
           'USE_ALL_GT': True,
           'USE_FLIPPED': True,
           'USE_GT': False,
           'WEIGHT_DECAY': 0.0005},
 'USE_GPU_NMS': True}
Loaded dataset `vg_alldata_val` for training
Set proposal method: gt
Preparing training data...
vg_alldata_val gt roidb loaded from /home/alex/faster-rcnn.pytorch/data/cache/vg_alldata_val_gt_roidb.pkl
done
4949 roidb entries
load checkpoint models/vgg16/vg/faster_rcnn_1_19_48611.pth
load model successfully!
~~~

**Comment by adityaarun1** [comment]
Looks fine to me. You may have to ask @jwyang for accuracy.

**Comment by kentaroy47** [comment]
Would be great if someone can upload a COCO trained model for the pytorch1.0 branch..
seems like there is a accuracy degradation there too.

**Comment by kangkang59812** [comment]
> Hi,
> 
> For the PyTorch-1.0 branch, I re-ran the code and it gives me 70.26% mAP when using VGG-16 and 74.90% using ResNet-101. These numbers are achieved using the default parameter values on a single GPU. These numbers are consistent with the ones reported by @jwyang in the readme. Here are the details:
> 
> VGG-16 (7th epoch)
> 
> ```
> VOC07 metric? Yes
> AP for aeroplane = 0.7071
> AP for bicycle = 0.7714
> AP for bird = 0.6922
> AP for boat = 0.5591
> AP for bottle = 0.5510
> AP for bus = 0.7913
> AP for car = 0.8458
> AP for cat = 0.8173
> AP for chair = 0.4943
> AP for cow = 0.7680
> AP for diningtable = 0.6090
> AP for dog = 0.8058
> AP for horse = 0.8283
> AP for motorbike = 0.7468
> AP for person = 0.7788
> AP for pottedplant = 0.4439
> AP for sheep = 0.7058
> AP for sofa = 0.6440
> AP for train = 0.7584
> AP for tvmonitor = 0.7345
> Mean AP = 0.7026
> ```
> 
> ResNet-101 (7th epoch)
> 
> ```
> VOC07 metric? Yes
> AP for aeroplane = 0.7883
> AP for bicycle = 0.8002
> AP for bird = 0.7752
> AP for boat = 0.6471
> AP for bottle = 0.6047
> AP for bus = 0.8135
> AP for car = 0.8585
> AP for cat = 0.8698
> AP for chair = 0.5443
> AP for cow = 0.8122
> AP for diningtable = 0.6728
> AP for dog = 0.8568
> AP for horse = 0.8484
> AP for motorbike = 0.7912
> AP for person = 0.7880
> AP for pottedplant = 0.4702
> AP for sheep = 0.7661
> AP for sofa = 0.7292
> AP for train = 0.7955
> AP for tvmonitor = 0.7487
> Mean AP = 0.7490
> ```

can you share the settings or the checkpoint? I still can't reproduce the result

**Comment by Anakindai** [comment]
> Command to run:
> 
> ```
> python3.6 test_net.py --dataset vg --net vgg16 \
>                       --checksession 1 --checkepoch 19 --checkpoint 48611 \
>                       --cuda
> ```
> 
> Print out:
> 
> ```
> Called with args:
> Namespace(cfg_file='cfgs/vgg16.yml', checkepoch=19, checkpoint=48611, checksession=1, class_agnostic=False, cuda=True, dataset='vg', large_scale=False, load_dir='models', mGPUs=False, net='vgg16', parallel_type=0, set_cfgs=None, vis=False)
> Using config:
> {'ANCHOR_RATIOS': [0.5, 1, 2],
>  'ANCHOR_SCALES': [4, 8, 16, 32],
>  'CROP_RESIZE_WITH_MAX_POOL': False,
>  'CUDA': False,
>  'DATA_DIR': '/home/alex/faster-rcnn.pytorch/data',
>  'DEDUP_BOXES': 0.0625,
>  'EPS': 1e-14,
>  'EXP_DIR': 'vgg16',
>  'FEAT_STRIDE': [16],
>  'GPU_ID': 0,
>  'MATLAB': 'matlab',
>  'MAX_NUM_GT_BOXES': 20,
>  'MOBILENET': {'DEPTH_MULTIPLIER': 1.0,
>                'FIXED_LAYERS': 5,
>                'REGU_DEPTH': False,
>                'WEIGHT_DECAY': 4e-05},
>  'PIXEL_MEANS': array([[[102.9801, 115.9465, 122.7717]]]),
>  'POOLING_MODE': 'align',
>  'POOLING_SIZE': 7,
>  'RESNET': {'FIXED_BLOCKS': 1, 'MAX_POOL': False},
>  'RNG_SEED': 3,
>  'ROOT_DIR': '/home/alex/faster-rcnn.pytorch',
>  'TEST': {'BBOX_REG': True,
>           'HAS_RPN': True,
>           'MAX_SIZE': 1000,
>           'MODE': 'nms',
>           'NMS': 0.3,
>           'PROPOSAL_METHOD': 'gt',
>           'RPN_MIN_SIZE': 16,
>           'RPN_NMS_THRESH': 0.7,
>           'RPN_POST_NMS_TOP_N': 300,
>           'RPN_PRE_NMS_TOP_N': 6000,
>           'RPN_TOP_N': 5000,
>           'SCALES': [600],
>           'SVM': False},
>  'TRAIN': {'ASPECT_GROUPING': False,
>            'BATCH_SIZE': 256,
>            'BBOX_INSIDE_WEIGHTS': [1.0, 1.0, 1.0, 1.0],
>            'BBOX_NORMALIZE_MEANS': [0.0, 0.0, 0.0, 0.0],
>            'BBOX_NORMALIZE_STDS': [0.1, 0.1, 0.2, 0.2],
>            'BBOX_NORMALIZE_TARGETS': True,
>            'BBOX_NORMALIZE_TARGETS_PRECOMPUTED': True,
>            'BBOX_REG': True,
>            'BBOX_THRESH': 0.5,
>            'BG_THRESH_HI': 0.5,
>            'BG_THRESH_LO': 0.0,
>            'BIAS_DECAY': False,
>            'BN_TRAIN': False,
>            'DISPLAY': 10,
>            'DOUBLE_BIAS': True,
>            'FG_FRACTION': 0.25,
>            'FG_THRESH': 0.5,
>            'GAMMA': 0.1,
>            'HAS_RPN': True,
>            'IMS_PER_BATCH': 1,
>            'LEARNING_RATE': 0.01,
>            'MAX_SIZE': 1000,
>            'MOMENTUM': 0.9,
>            'PROPOSAL_METHOD': 'gt',
>            'RPN_BATCHSIZE': 256,
>            'RPN_BBOX_INSIDE_WEIGHTS': [1.0, 1.0, 1.0, 1.0],
>            'RPN_CLOBBER_POSITIVES': False,
>            'RPN_FG_FRACTION': 0.5,
>            'RPN_MIN_SIZE': 8,
>            'RPN_NEGATIVE_OVERLAP': 0.3,
>            'RPN_NMS_THRESH': 0.7,
>            'RPN_POSITIVE_OVERLAP': 0.7,
>            'RPN_POSITIVE_WEIGHT': -1.0,
>            'RPN_POST_NMS_TOP_N': 2000,
>            'RPN_PRE_NMS_TOP_N': 12000,
>            'SCALES': [600],
>            'SNAPSHOT_ITERS': 5000,
>            'SNAPSHOT_KEPT': 3,
>            'SNAPSHOT_PREFIX': 'res101_faster_rcnn',
>            'STEPSIZE': [30000],
>            'SUMMARY_INTERVAL': 180,
>            'TRIM_HEIGHT': 600,
>            'TRIM_WIDTH': 600,
>            'TRUNCATED': False,
>            'USE_ALL_GT': True,
>            'USE_FLIPPED': True,
>            'USE_GT': False,
>            'WEIGHT_DECAY': 0.0005},
>  'USE_GPU_NMS': True}
> Loaded dataset `vg_alldata_val` for training
> Set proposal method: gt
> Preparing training data...
> vg_alldata_val gt roidb loaded from /home/alex/faster-rcnn.pytorch/data/cache/vg_alldata_val_gt_roidb.pkl
> done
> 4949 roidb entries
> load checkpoint models/vgg16/vg/faster_rcnn_1_19_48611.pth
> load model successfully!
> ```

After run the preprocessing script provided by the bottom-up-attention, need I change the generated files' path? Why my Print out "99 roidb entries" rather that much?
